testdata/p4_16_samples_outputs/p416-type-use3.p4.p4info.txt

```diff
@@ -0,0 +1,216 @@
+    match_type: EXACT
+    type_name {
+      name: "EthT_t"
+    }
+  }
+  match_fields {
+    id: 2
+    name: "hdr.ethernet.dstAddr"
+    bitwidth: 48
+    match_type: EXACT
+  }
+  match_fields {
+    id: 3
+    name: "hdr.custom.e"
+    bitwidth: 8
+    match_type: EXACT
+  }
+  match_fields {
+    id: 4
+    name: "hdr.custom.ed"
+    bitwidth: 8
+    match_type: EXACT
+  }
+  match_fields {
+    id: 5
+    name: "hdr.custom.et"
+    bitwidth: 8
+    match_type: EXACT
+  }
+  match_fields {
+    id: 6
+    name: "hdr.custom.edd"
+    bitwidth: 8
+    match_type: EXACT
+  }
+  match_fields {
+    id: 7
+    name: "hdr.custom.eddt"
+    bitwidth: 8
+    match_type: EXACT
+  }
+  match_fields {
+    id: 8
+    name: "hdr.custom.edtd"
+    bitwidth: 8
+    match_type: EXACT
+  }
+  match_fields {
+    id: 9
+    name: "hdr.custom.edtt"
+    bitwidth: 8
+    match_type: EXACT
+  }
+  match_fields {
+    id: 10
+    name: "hdr.custom.etdd"
+    bitwidth: 8
+    match_type: EXACT
+  }
+  match_fields {
+    id: 11
+    name: "hdr.custom.etdt"
+    bitwidth: 8
+    match_type: EXACT
+    type_name {
+      name: "CustomTDT_t"
+    }
+  }
+  match_fields {
+    id: 12
+    name: "hdr.custom.ettd"
+    bitwidth: 8
+    match_type: EXACT
+  }
+  match_fields {
+    id: 13
+    name: "hdr.custom.ettt"
+    bitwidth: 8
+    match_type: EXACT
+    type_name {
+      name: "CustomTTT_t"
+    }
+  }
+  action_refs {
+    id: 16829338
+  }
+  action_refs {
+    id: 16797750
+  }
+  action_refs {
+    id: 16815499
+  }
+  size: 1024
+}
+actions {
+  preamble {
+    id: 16815499
+    name: "my_drop"
+    alias: "my_drop"
+  }
+}
+actions {
+  preamble {
+    id: 16829338
+    name: "ingress.set_output"
+    alias: "set_output"
+  }
+  params {
+    id: 1
+    name: "out_port"
+    bitwidth: 9
+  }
+}
+actions {
+  preamble {
+    id: 16797750
+    name: "ingress.set_headers"
+    alias: "set_headers"
+  }
+  params {
+    id: 1
+    name: "e"
+    bitwidth: 8
+  }
+  params {
+    id: 2
+    name: "ed"
+    bitwidth: 8
+  }
+  params {
+    id: 3
+    name: "et"
+    bitwidth: 8
+  }
+  params {
+    id: 4
+    name: "edd"
+    bitwidth: 8
+  }
+  params {
+    id: 5
+    name: "edt"
+    bitwidth: 8
+  }
+  params {
+    id: 6
+    name: "etd"
+    bitwidth: 8
+  }
+  params {
+    id: 7
+    name: "ett"
+    bitwidth: 8
+  }
+  params {
+    id: 8
+    name: "eddd"
+    bitwidth: 8
+  }
+  params {
+    id: 9
+    name: "eddt"
+    bitwidth: 8
+  }
+  params {
+    id: 10
+    name: "edtd"
+    bitwidth: 8
+  }
+  params {
+    id: 11
+    name: "edtt"
+    bitwidth: 8
+  }
+  params {
+    id: 12
+    name: "etdd"
+    bitwidth: 8
+  }
+  params {
+    id: 13
+    name: "etdt"
+    bitwidth: 8
+    type_name {
+      name: "CustomTDT_t"
+    }
+  }
+  params {
+    id: 14
+    name: "ettd"
+    bitwidth: 8
+  }
+  params {
+    id: 15
+    name: "ettt"
+    bitwidth: 8
+    type_name {
+      name: "CustomTTT_t"
+    }
+  }
+}
+type_info {
```

Comment:
Any type name given in a `type_name` earlier in the P4Info file should have a message describing it in the `type_info` body here.

Reply:
If you notice the p4info.txt for `p4rt_digest_complex.p4`, the file contains `new_types` in the `type_info`.  The p4c p4runtime API generation supports only Digest and Register for converting/translating types.  This is why a table MatchField or Action Param do not have new_types - the code changes to support these types will need some thought and I may punt to a subsequent PR.

Reply:
Yes, I did notice that `type_info` is not always empty in all of these examples.  As I mentioned in another comment, I do not necessarily think that these P4Info files must be perfect before these changes can be merged in, as long as there is an issue to remind us to do the additional work to correct them later.

Reply:
Andy is correct. P4TypeInfo cannot be empty when there are some set `type_name` fields in your P4Info...